File: src/solver/subroutines/solveMiddleEdges.ts
import Cube from '../../engine/Cube'
import { cubesShareFace, positionMap } from '../../utils'
import RubiksCubeSolver from '../RubiksCubeSolver'

export default async function solveMiddleEdges(solver: RubiksCubeSolver) {
  const bottomEdge = solver.rubiks.cubes.find((cube) => cube.isInBottomLayer && !cube.hasFace('W'))

  if (bottomEdge) {
    await solveBottomEdge(solver, bottomEdge)
    return
  }

  const unsolvedMiddleEdge = solver.rubiks.cubes
    .filter((cube) => cube.isInXMidLayer)
    .find((cube) => {
      return (
        (cube.orientation.front &&
          !cubesShareFace('front', cube, solver.cubeMap.get(positionMap[17]))) ||
        (cube.orientation.left &&
          !cubesShareFace('left', cube, solver.cubeMap.get(positionMap[13]))) ||
        (cube.orientation.back &&
          !cubesShareFace('back', cube, solver.cubeMap.get(positionMap[11]))) ||
        (cube.orientation.right &&
          !cubesShareFace('right', cube, solver.cubeMap.get(positionMap[15])))
      )
    })!

  if (unsolvedMiddleEdge.isInLeftLayer) {
    if (unsolvedMiddleEdge.isInBackLayer) {
      await solver.do('XCCW')
    }
    await solver.do(
      'rotateBottomCCW',
      'rotateLeftCCW',
      'rotateBottomCW',
      'rotateLeftCW',
      'rotateBottomCW',
      'rotateFrontCCW',
      'rotateBottomCCW',
      'rotateFrontCW'
    )
  }
  if (unsolvedMiddleEdge.isInRightLayer) {
    if (unsolvedMiddleEdge.isInBackLayer) {
      await solver.do('XCW')
    }
    await solver.do(
      'rotateBottomCW',
      'rotateRightCCW',
      'rotateBottomCCW',
      'rotateRightCW',
      'rotateBottomCCW',
      'rotateFrontCW',
      'rotateBottomCW',
      'rotateFrontCCW'
    )
  }

  await solveBottomEdge(solver, unsolvedMiddleEdge)
  return
}

async function solveBottomEdge(solver: RubiksCubeSolver, bottomEdge: Cube) {
  if (bottomEdge.isInBackLayer) {
    await solver.do('XCCW', 'XCCW')
  } else if (bottomEdge.isInLeftLayer) {
    await solver.do('XCCW')
  } else if (bottomEdge.isInRightLayer) {
    await solver.do('XCW')
  }

  while (!cubesShareFace('front', bottomEdge, solver.cubeMap.get(positionMap[17]))) {
    await solver.do('rotateBottomCW', 'XCCW')
  }

  if (bottomEdge.orientation.bottom === solver.cubeMap.get(positionMap[13])?.orientation.left) {
    await solver.do(
      'rotateBottomCCW',
      'rotateLeftCCW',
      'rotateBottomCW',
      'rotateLeftCW',
      'rotateBottomCW',
      'rotateFrontCCW',
      'rotateBottomCCW',
      'rotateFrontCW'
    )
  } else {
    await solver.do(
      'rotateBottomCW',
      'rotateRightCCW',
      'rotateBottomCCW',
      'rotateRightCW',
      'rotateBottomCCW',
      'rotateFrontCW',
      'rotateBottomCW',
      'rotateFrontCCW'
    )
  }
  return
}

File: android/core/src/main/kotlin/com/mikeb/simplepuzzlecube/solver/subroutines/solveMiddleEdges.kt
// Port of ../src/solver/subroutines/solveMiddleEdges.ts.
package com.mikeb.simplepuzzlecube.solver.subroutines

import com.mikeb.simplepuzzlecube.cubesShareFace
import com.mikeb.simplepuzzlecube.engine.Cube
import com.mikeb.simplepuzzlecube.engine.Face
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateBottomCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateBottomCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateFrontCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateFrontCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateLeftCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateLeftCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCCW
import com.mikeb.simplepuzzlecube.engine.LayerMove.rotateRightCW
import com.mikeb.simplepuzzlecube.engine.OrientationKey
import com.mikeb.simplepuzzlecube.engine.Rotation.XCCW
import com.mikeb.simplepuzzlecube.engine.Rotation.XCW
import com.mikeb.simplepuzzlecube.solver.RubiksCubeSolver

suspend fun solveMiddleEdges(solver: RubiksCubeSolver) {
    val bottomEdge = solver.rubiks.cubes.find { cube ->
        cube.isInBottomLayer && !cube.hasFace(Face.W)
    }

    if (bottomEdge != null) {
        solveBottomEdge(solver, bottomEdge)
        return
    }

    val unsolvedMiddleEdge = solver.rubiks.cubes
        .filter { cube -> cube.isInXMidLayer }
        .find { cube ->
            (cube.orientation.front != null && !cubesShareFace(OrientationKey.front, cube, solver.fetchPosition(17))) ||
                (cube.orientation.left != null && !cubesShareFace(OrientationKey.left, cube, solver.fetchPosition(13))) ||
                (cube.orientation.back != null && !cubesShareFace(OrientationKey.back, cube, solver.fetchPosition(11))) ||
                (cube.orientation.right != null && !cubesShareFace(OrientationKey.right, cube, solver.fetchPosition(15)))
        }!!

    if (unsolvedMiddleEdge.isInLeftLayer) {
        if (unsolvedMiddleEdge.isInBackLayer) {
            solver.`do`(XCCW)
        }
        solver.`do`(
            rotateBottomCCW,
            rotateLeftCCW,
            rotateBottomCW,
            rotateLeftCW,
            rotateBottomCW,
            rotateFrontCCW,
            rotateBottomCCW,
            rotateFrontCW
        )
    }
    if (unsolvedMiddleEdge.isInRightLayer) {
        if (unsolvedMiddleEdge.isInBackLayer) {
            solver.`do`(XCW)
        }
        solver.`do`(
            rotateBottomCW,
            rotateRightCCW,
            rotateBottomCCW,
            rotateRightCW,
            rotateBottomCCW,
            rotateFrontCW,
            rotateBottomCW,
            rotateFrontCCW
        )
    }

    solveBottomEdge(solver, unsolvedMiddleEdge)
    return
}

private suspend fun solveBottomEdge(solver: RubiksCubeSolver, bottomEdge: Cube) {
    if (bottomEdge.isInBackLayer) {
        solver.`do`(XCCW, XCCW)
    } else if (bottomEdge.isInLeftLayer) {
        solver.`do`(XCCW)
    } else if (bottomEdge.isInRightLayer) {
        solver.`do`(XCW)
    }

    while (!cubesShareFace(OrientationKey.front, bottomEdge, solver.fetchPosition(17))) {
        solver.`do`(rotateBottomCW, XCCW)
    }

    if (bottomEdge.orientation.bottom == solver.fetchPosition(13)?.orientation?.left) {
        solver.`do`(
            rotateBottomCCW,
            rotateLeftCCW,
            rotateBottomCW,
            rotateLeftCW,
            rotateBottomCW,
            rotateFrontCCW,
            rotateBottomCCW,
            rotateFrontCW
        )
    } else {
        solver.`do`(
            rotateBottomCW,
            rotateRightCCW,
            rotateBottomCCW,
            rotateRightCW,
            rotateBottomCCW,
            rotateFrontCW,
            rotateBottomCW,
            rotateFrontCCW
        )
    }
    return
}
